pkg/convert/jfr/parser.go

```diff
@@ -14,155 +14,224 @@ import (
+type Tags struct {
+	Contexts map[int64]map[int64]int64 `json:"contexts"`
+	Strings  map[int64]string          `json:"strings"`
+}
+
+func ParseJFR(ctx context.Context, r io.Reader, s storage.Putter, pi *storage.PutInput, tags *Tags) (err error) {
 	chunks, err := parser.Parse(r)
 	if err != nil {
 		return fmt.Errorf("unable to parse JFR format: %w", err)
 	}
 	for _, c := range chunks {
-		if pErr := parse(ctx, c, s, pi); pErr != nil {
+		if pErr := parse(ctx, c, s, pi, tags); pErr != nil {
 			err = multierror.Append(err, pErr)
 		}
 	}
 	pi.Val = nil
 	return err
 }
+func resolveTags(contextId int64, tags *Tags) map[string]string {
+	res := make(map[string]string)
+	if contextId == 0 {
+		return res
+	}
+	var ctx map[int64]int64
+	var ok bool
+	if ctx, ok = tags.Contexts[contextId]; !ok {
+		return nil
+	}
+	for k, v := range ctx {
+		var ks string
+		var vs string
+		if ks, ok = tags.Strings[k]; !ok {
+			continue
+		}
+		if vs, ok = tags.Strings[v]; !ok {
+			continue
+		}
+		res[ks] = vs
+	}
+	return res
+}
 
-func parse(ctx context.Context, c parser.Chunk, s storage.Putter, pi *storage.PutInput) (err error) {
+func parse(ctx context.Context, c parser.Chunk, s storage.Putter, pi *storage.PutInput, tags *Tags) (err error) {
 	var event, alloc, lock string
-	cpu := tree.New()
-	wall := tree.New()
-	inTLABObjects := tree.New()
-	inTLABBytes := tree.New()
-	outTLABObjects := tree.New()
-	outTLABBytes := tree.New()
-	lockSamples := tree.New()
-	lockDuration := tree.New()
 	for _, e := range c.Events {
 		switch e.(type) {
-		case *parser.ExecutionSample:
-			es := e.(*parser.ExecutionSample)
-			if fs := frames(es.StackTrace); fs != nil {
-				if es.State.Name == "STATE_RUNNABLE" {
-					cpu.InsertStackString(fs, 1)
-				}
-				wall.InsertStackString(fs, 1)
-			}
-		case *parser.ObjectAllocationInNewTLAB:
-			oa := e.(*parser.ObjectAllocationInNewTLAB)
-			if fs := frames(oa.StackTrace); fs != nil {
-				inTLABObjects.InsertStackString(fs, 1)
-				inTLABBytes.InsertStackString(fs, uint64(oa.TLABSize))
-			}
-		case *parser.ObjectAllocationOutsideTLAB:
-			oa := e.(*parser.ObjectAllocationOutsideTLAB)
-			if fs := frames(oa.StackTrace); fs != nil {
-				outTLABObjects.InsertStackString(fs, 1)
-				outTLABBytes.InsertStackString(fs, uint64(oa.AllocationSize))
-			}
-		case *parser.JavaMonitorEnter:
-			jme := e.(*parser.JavaMonitorEnter)
-			if fs := frames(jme.StackTrace); fs != nil {
-				lockSamples.InsertStackString(fs, 1)
-				lockDuration.InsertStackString(fs, uint64(jme.Duration))
-			}
-		case *parser.ThreadPark:
-			tp := e.(*parser.ThreadPark)
-			if fs := frames(tp.StackTrace); fs != nil {
-				lockSamples.InsertStackString(fs, 1)
-				lockDuration.InsertStackString(fs, uint64(tp.Duration))
-			}
 		case *parser.ActiveSetting:
 			as := e.(*parser.ActiveSetting)
 			switch as.Name {
 			case "event":
-				event = as.Value
+				event = as.Value // this can be passed from the agent as upload param
```

Comment:
Also, this will only work if there's one type per chunk. Can we guarantee that that's always true? (I don't know too much about JFR internals)

I think if we can guarantee it, then this code is totally fine, we just need to remove this comment.

Reply:
I removed the comment and left the things as they were before the PR